app/controllers/application_controller.rb

```diff
@@ -1,3 +1,35 @@
 class ApplicationController < ActionController::Base
   protect_from_forgery with: :exception
+
+  helper_method :current_user, :logged_in?, :logged_out?, :require_user, :require_logged_out, :access_denied
+
+  def current_user
+    @current_user ||= User.find(session[:user_id]) if session[:user_id]
+  end
+
+  def logged_in?
+    !!current_user
+  end
+
+  def logged_out?
+    !logged_in?
+  end
+
+  def require_user
+    access_denied if logged_out?
+  end
+
+  def require_logged_out
+    access_denied if logged_in? 
+  end
+
+  def access_denied
+    if logged_out?
+      flash[:error] = "You must be logged in to do that"
+      redirect_to root_path
+    else
+      flash[:error] = "You must be logged out to do that"
```

Comment:
I am not sure about this error message. Its rare that we expect user to be logged out.

Reply:
this method is used as a `before_action` in the Users and Sessions controller for the `new` method in both.  A logged in user can't create a new user.  Nor can they visit the session new path to log in.  Granted this would only occur if a user went to the trouble of visting that URL directly.